### PR TITLE
enhance: retrieve output fields after local reduce

### PR DESCRIPTION
**Pull request by longjiquan**
issue: #31822

### CODE REVIEW

**Comment by longjiquan** [comment]
rerun ut

**Comment by longjiquan** [comment]
rerun ut

**Comment by longjiquan** [comment]
rerun ut

**Comment by longjiquan** [comment]
rerun ut

**Comment by longjiquan** [comment]
rerun ut

**Comment by longjiquan** [comment]
rerun ut

File: pkg/util/typeutil/schema.go [hunk 1]
```diff
@@ -568,6 +568,121 @@ func PrepareResultFieldData(sample []*schemapb.FieldData, topK int64) []*schemap
 	return result
 }
 
+func ReserveSize(dst, src *schemapb.FieldData, size int64) {
```
Comment (bigsheeper) [inline pkg/util/typeutil/schema.go:571]: no usage?

File: internal/querynodev2/segments/segment.go [hunk 2]
```diff
@@ -579,6 +584,65 @@ func (s *LocalSegment) Retrieve(ctx context.Context, plan *RetrievePlan) (*segco
...
+
+	var retrieveResult RetrieveResult
+	var status C.CStatus
+	GetSQPool().Submit(func() (any, error) {
+		tr := timerecord.NewTimeRecorder("cgoRetrieveByOffsets")
+		status = C.RetrieveByOffsets(traceCtx,
+			s.ptr,
+			plan.cRetrievePlan,
+			&retrieveResult.cRetrieveResult,
+			(*C.int64_t)(unsafe.Pointer(&offsets[0])),
```
Comment (bigsheeper) [inline internal/querynodev2/segments/segment.go:620]: maybe need to check the length of `offsets` to prevent panic

File: internal/querynodev2/segments/result.go [hunk 3]
```diff
@@ -524,25 +538,80 @@ func MergeSegcoreRetrieveResults(ctx context.Context, retrieveResults []*segcore
...
+		defer span2.End()
+		segmentResults := make([]*segcorepb.RetrieveResults, len(validRetrieveResults))
+		eg, ctx := errgroup.WithContext(ctx)
+		for i, offsets := range selectedOffsets {
+			if len(offsets) == 0 {
+				log.Ctx(ctx).Debug("skip empty retrieve results", zap.Int64("segment", validSegments[i].ID()))
+				continue
+			}
+			idx, theOffsets := i, offsets
+			eg.Go(func() error {
```
Comment (bigsheeper) [inline internal/querynodev2/segments/result.go:581]: use thread pool instead

File: pkg/util/typeutil/schema.go (hunk 1 above)
Comment (longjiquan) [reply pkg/util/typeutil/schema.go:571]: Yeah, I'll delete this function.

File: internal/querynodev2/segments/segment.go [hunk 4]
```diff
@@ -579,6 +584,65 @@ func (s *LocalSegment) Retrieve(ctx context.Context, plan *RetrievePlan) (*segco
...
 }
 
+func (s *LocalSegment) RetrieveByOffsets(ctx context.Context, plan *RetrievePlan, offsets []int64) (*segcorepb.RetrieveResults, error) {
+	if !s.ptrLock.RLockIf(state.IsNotReleased) {
+		// TODO: check if the segment is readable but not released. too many related logic need to be refactor.
+		return nil, merr.WrapErrSegmentNotLoaded(s.ID(), "segment released")
+	}
+	defer s.ptrLock.RUnlock()
+
+	if s.ptr == nil {
```
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/querynodev2/segments/segment.go:594]: Just similar to LocalSegment.Retrieve, I think it's OK to keep this check in case of any assumption is broken (for example, s.ptr will be nil).

File: internal/querynodev2/segments/segment.go (hunk 2 above)
Comment (longjiquan) [reply internal/querynodev2/segments/segment.go:620]: OK.

File: internal/querynodev2/segments/result.go (hunk 3 above)
Comment (longjiquan) [reply internal/querynodev2/segments/result.go:581]: Should we use which pool to schedule this?

Comment (longjiquan) [reply internal/querynodev2/segments/result.go:581]: Btw, the inner RetrieveByOffsets was already scheduled by the query thread pool.

Comment (longjiquan) [reply internal/querynodev2/segments/result.go:581]: I don't find we take any concurrent control on `retrieveOnSegments`, neither.

File: internal/querynodev2/segments/retrieve.go [hunk 5]
```diff
@@ -38,13 +38,19 @@ import (
...
+		segment Segment
+	}
 	var (
-		resultCh = make(chan *segcorepb.RetrieveResults, len(segments))
+		resultCh = make(chan segmentResult, len(segments))
 		errs     = make([]error, len(segments))
 		wg       sync.WaitGroup
 	)
 
+	plan.ignoreNonPk = len(segments) > 1 && plan.ShouldIgnoreNonPk()
```
Comment (bigsheeper) [inline internal/querynodev2/segments/retrieve.go:52]: don't ignore non pk when user queries without pagination

File: internal/core/src/segcore/SegmentInterface.cpp [hunk 6]
```diff
@@ -173,7 +202,23 @@ SegmentInternalInterface::Retrieve(const query::RetrievePlan* plan,
                 }
             }
         }
+        if (!ignore_non_pk) {
+            // when ignore_non_pk is false, it indicates two situations:
```
Comment (bigsheeper) [inline internal/core/src/segcore/SegmentInterface.cpp:206]: modify the comment as well

File: internal/core/src/segcore/SegmentInterface.h [hunk 7]
```diff
@@ -279,6 +301,15 @@ class SegmentInternalInterface : public SegmentInterface {
...
                bool false_filtered_out) const = 0;
 
+    void
+    FillTargetEntry(
+        const query::RetrievePlan* plan,
+        const std::unique_ptr<proto::segcore::RetrieveResults>& results,
+        const int64_t* offsets,
+        int64_t size,
+        bool ignore_non_pk,
+        bool pk_in_ids) const;
```
Comment (bigsheeper) [inline internal/core/src/segcore/SegmentInterface.h:311]: confused name

File: internal/querynodev2/segments/segment.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (bigsheeper) [reply internal/querynodev2/segments/segment.go:594]: LocalSegment.Retrieve didn't make this check. Before we have a common way to check if segment is deleted, I think this check is unnecessary.

**Comment by longjiquan** [comment]
comments fixed at https://github.com/milvus-io/milvus/pull/32346/commits/36efedc54998a5708b5dbf8654effa57bae05ebb, cc @bigsheeper

File: internal/querynodev2/segments/result.go [hunk 8]
```diff
@@ -524,25 +538,80 @@ func MergeSegcoreRetrieveResults(ctx context.Context, retrieveResults []*segcore
...
+		defer span3.End()
+		cursors = make([]int64, len(segmentResults))
+		for _, sel := range selected {
+			retSize += typeutil.AppendFieldData(ret.FieldsData, segmentResults[sel].GetFieldsData(), cursors[sel])
+			cursors[sel]++
+		}
+	}
+
+	// limit retrieve result to avoid oom
+	if retSize > maxOutputSize {
```
Comment (czs007) [inline internal/querynodev2/segments/result.go:611]: revert this change

File: internal/querynodev2/segments/result.go (hunk 3 above)
Comment (czs007) [inline internal/querynodev2/segments/result.go:581]: It is advisable to avoid creating many  goroutines here.

File: internal/core/src/segcore/segment_c.cpp [hunk 9]
```diff
@@ -130,18 +130,51 @@ Retrieve(CTraceContext c_trace,
...
-        auto span = milvus::tracer::StartSpan("SegCoreRetrieve", &trace_ctx);
-        milvus::tracer::SetRootSpan(span);
+        milvus::tracer::AutoSpan span("SegCoreRetrieve", &trace_ctx, true);
 
-        auto retrieve_result = segment->Retrieve(plan, timestamp, limit_size);
+        auto retrieve_result =
+            segment->Retrieve(plan, timestamp, limit_size, ignore_non_pk);
+
+        auto size = retrieve_result->ByteSizeLong();
+        void* buffer = malloc(size);
```
Comment (czs007) [inline internal/core/src/segcore/segment_c.cpp:148]: make use of RAII to ensure exception safety. you can modify this in another pr.

**Comment by longjiquan** [comment]
comments fixed at https://github.com/milvus-io/milvus/pull/32346/commits/5e50e7e27c8534df9e917190e03d4373799b96d6, cc @czs007
